scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordser/ProductOrderedBuf.scala

```diff
@@ -0,0 +1,122 @@
+      typeOf[Product5[Any, Any, Any, Any, Any]],
+      typeOf[Product6[Any, Any, Any, Any, Any, Any]],
+      typeOf[Product7[Any, Any, Any, Any, Any, Any, Any]],
+      typeOf[Product8[Any, Any, Any, Any, Any, Any, Any, Any]],
+      typeOf[Product9[Any, Any, Any, Any, Any, Any, Any, Any, Any]],
+      typeOf[Product10[Any, Any, Any, Any, Any, Any, Any, Any, Any, Any]],
+      typeOf[Product11[Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any]],
+      typeOf[Product12[Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any]],
+      typeOf[Product13[Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any]],
+      typeOf[Product14[Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any]],
+      typeOf[Product15[Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any]],
+      typeOf[Product16[Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any]],
+      typeOf[Product17[Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any]],
+      typeOf[Product18[Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any]],
+      typeOf[Product19[Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any]],
+      typeOf[Product20[Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any]],
+      typeOf[Product21[Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any]],
+      typeOf[Product22[Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any, Any]])
+
+    def validType(curType: Type): Boolean = {
+      validTypes.find{ t => curType <:< t }.isDefined
+    }
+
+    def symbolFor(subType: Type): Type = {
+      val superType = validTypes.find{ t => subType.erasure <:< t }.get
+      subType.baseType(superType.typeSymbol)
+    }
+
+    val pf: PartialFunction[c.Type, TreeOrderedBuf[c.type]] = {
+      case tpe if validType(tpe.erasure) => ProductOrderedBuf(c)(buildDispatcher, tpe, symbolFor(tpe))
+    }
+    pf
+  }
+
+  def apply(c: Context)(buildDispatcher: => PartialFunction[c.Type, TreeOrderedBuf[c.type]], originalType: c.Type, outerType: c.Type): TreeOrderedBuf[c.type] = {
+    import c.universe._
+    def freshT(id: String) = newTermName(c.fresh(id))
+
+    val dispatcher = buildDispatcher
+    val elementData: List[(c.universe.Type, TermName, TreeOrderedBuf[c.type])] =
+      outerType
+        .declarations
+        .collect { case m: MethodSymbol => m }
+        .filter(m => m.name.toTermName.toString.startsWith("_"))
+        .map { accessorMethod =>
+          val fieldType = accessorMethod.returnType.asSeenFrom(outerType, outerType.typeSymbol.asClass)
+          val b: TreeOrderedBuf[c.type] = dispatcher(fieldType)
+          (fieldType, accessorMethod.name.toTermName, b)
+        }.toList
+
+    new TreeOrderedBuf[c.type] {
```

Comment:
this looks almost identical to the case class code. Can't we share code here and just pass a Tree to instantiate the new object?